In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
#For linear Regression using the statsmodels package
import statsmodels.api as stm

C:\Users\lenovo\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:49: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  data_klasses = (pandas.Series, pandas.DataFrame, pandas.Panel)


In [3]:
#For linear Regression using the sklearn package
import scipy.stats as stats
import sklearn

In [4]:
os.chdir(r"C:\Users\lenovo\Desktop\Praxis\Practice_doc\Python")

In [5]:
cars = pd.read_csv("cars.csv")

In [6]:
cars.columns

Index(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight',
       'Acceleration', 'Model', 'Origin'],
      dtype='object')

In [7]:
X = cars[['Weight','Displacement']]
Y = cars['MPG']

### Fitting Linear Regression model

In [8]:
X = stm.add_constant(X)

C:\Users\lenovo\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [9]:
#Ordinary Least Squared (OLS) Regression
model = stm.OLS(Y,X).fit()  
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    MPG   R-squared:                       0.627
Model:                            OLS   Adj. R-squared:                  0.625
Method:                 Least Squares   F-statistic:                     338.8
Date:                Fri, 10 Jan 2020   Prob (F-statistic):           4.89e-87
Time:                        19:19:36   Log-Likelihood:                -1239.5
No. Observations:                 406   AIC:                             2485.
Df Residuals:                     403   BIC:                             2497.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           43.5806      1.374     31.716      0.000      40.879      46.282
Weight          -0.0057      0.001     -6.810      0.000      -0.007      -0.004
Displacement    -0.0184      0.007     -2.725      0.007      -0.032      -0.005
==============================================================================
Omnibus:                       95.903   Durbin-Watson:                   1.024
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              634.672
Skew:                          -0.815   Prob(JB):                    1.52e-138
Kurtosis:                       8.904   Cond. No.                     1.67e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.67e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [10]:
# To extract the standard error from the fitted model
model.params[0]

43.58055285352251

In [11]:
# To extract the standard error from the fitted model for a specific predictor
model.bse.const

1.374100806290394

In [12]:
model.bse.Weight

0.0008353566889100266

## Using Bootstrapping

In [13]:
# Step 1 : For Bootstrapping take a sample from original dataframe with replacement False

sample1 = cars.sample(200, replace = False)

In [14]:
# List to store all the standard error
s_e_const = []
s_e_wt = []
s_e_disp = []

for i in range(10):
    
    # Step 2 : Take multiple sample from above sample with replacment true
    df = sample1.sample(200, replace = True)
    
    # Define predictors and target variable
    X = df[['Weight','Displacement']]
    Y = df['MPG'] 
    
    # Fitting the linear regression model
    X = stm.add_constant(X)
    model = stm.OLS(Y,X).fit()
    
    # Extracting
    const = model.params[0]
    s_e_const.append(const)
    Weight = model.params[1]
    s_e_wt.append(Weight)
    Displacement = model.params[2]
    s_e_disp.append(Displacement)

In [15]:
np.std(s_e_const)/np.sqrt(200)

0.09814274572055516

In [16]:
import statistics
statistics.stdev(s_e_const)/np.sqrt(200)

0.10345153743323247

In [19]:
round(statistics.stdev(s_e_wt)/np.sqrt(200),6)

5.2e-05

In [18]:
statistics.stdev(s_e_disp)

0.006440493750731655

### Conclusion :

Standard error for entire dataframe :
    const           1.374101
    Weight          0.000835
    Displacement    0.006744

Standard error after using bootstrapping :
    const           1.965156
    Weight          0.001166
    Displacement    0.009141